### PR TITLE
Partially fixes #294 - Added form validation in frontend using JS

### LINKED ISSUE #294: Add error info to specific field and include details
When a user enters data that is not correct entered and hits "Upload" a general error message appears. Please make this easier to understand and add more information.

1. Highlight the box that was erroneous in red and add info that it is erroneous
2. Keep the information that was previously entered in all fields
3. Provide a helper text how the information should be correctly formatted

screenshot from 2017-10-22 17-42-54

See screenshot above. The user does not know what is the correct format. There is no info anywhere what the correct format is like.

### PR DESCRIPTION
**Pull request by djmgit**
Partially solves issue #294, added JS form validation. Added
field specific errors.

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by djmgit** [comment]
@S2606 @gabru-md @jajodiaraghav I have moved the form validation to frontend using JS. I have tried to cover almost all types of errors. Now when a error occurs the page does not get refreshed. The entered data persists. Now we do not require form validation on server side. Also we do not have to require on passing error messages using flash between html and flask. Please review :)

**Comment by jajodiaraghav** [comment]
Great work @djmgit 
My thoughts:
1. I think we already have an "error" class, we do not have to add a new CSS class.
2. I think the `default.csv.png` file could be modified to be more meaningful.
3. Even after merging this, we should not remove server side validation.

I would like to know what others think. Rest looks good to me.
Great work.

File: app/static/uploads/default.png.csv [hunk 1]
```diff
@@ -0,0 +1,2 @@
+a,b,c,d
```
Comment (gabru-md) [inline app/static/uploads/default.png.csv]: Make this file a bit more meaningful like adding real data for eg 
```sh
Mario Behling,@mariobehling,Owner,FOSSASIA
Manish Devgan,@gabru-md,Badgeyay,FOSSASIA
Raghav Jajodia,@jajodiaraghav,Badgeyay,FOSSASIA 
etc
```

File: app/templates/index.html [hunk 2]
```diff
@@ -45,19 +45,24 @@
...
 					<textarea class="form-control" id="textArea" name="csv"></textarea>
-			  	</div>
-			  	<div class="form-group">
-						<div class="text-center guide">
-                 Check out our <a href="/guide" target="_blank">User Input Guide</a>
-						</div>
+				</div>
+				<div id="error" class="no-data-error"> Please upload a csv file or enter data! </div>
+				<div id="error" class="csv-error"> Please fill data in proper format
+					<div class="text-center guide">
```
Comment (S2606) [inline app/templates/index.html]: @djmgit  How about changing `Please fill data in proper format` it to `Do follow the data guidlines. For more info:` ?

**Comment by gabru-md** [comment]
@djmgit Thanks for working on front-end verification. It is great.
Just make the changes as requested.
@jajodiaraghav I guess it would be okay to remove Server side verification as it is being handled on client side. If you could please let us know why are you in favor of keeping the server side verification. Maybe your explanation can help us understand your point better. :+1:

File: app/templates/index.html (hunk 2 above)
Comment (S2606) [reply app/templates/index.html]: Else Nice Work @djmgit!!

**Comment by jajodiaraghav** [comment]
@gabru-md My thoughts on this:
1. JavaScript code can be disabled or altered on a web browser.
2. We might in future add a feature where data is fetched from an API, in that can client side validation won't work (not sure)

**Comment by gabru-md** [comment]
Thank you @jajodiaraghav for pointing out these vulnerabilities.
That was convincing and hence server side verification must not be removed as of now.
Cheers!

**Comment by ParthS007** [comment]
@djmgit You can do the changes after UI is changed otherwise you have to write code again as their are many changes in form

**Comment by djmgit** [comment]
@ParthS007 You are right, I will wait till #309 is merged :)
@jajodiaraghav Viewing of JS code via browser is really an issue. As of your second point, if we fetch data from API or post data to API then we will have to use further more number of validation functions which will validate the data being posted. In future if we start using AJAX, or think of separating the frontend from backend  (ie change the back end to a REST API) then we will be dependent on JS even more and in that case your first pint will become a greater issue. For browsers with disabled JS, I am not sure what is the solution but then most of the applications now a days are decoupled and our entirely dependent on front end JS libraries like react or angular. As of issue regarding viewing JS code on browser, we have to make sure we that our JS code does not expose any important data (like any sort of keys). We can always use some build tool (via npm) which will use some libraries like uglifyjs which will minify and mangle our JS code so that no one can make sense of what they are seeing.

**Comment by jajodiaraghav** [comment]
@djmgit Now that #309 is merged, we need to update this so that we can finally have client side validation

**Comment by djmgit** [comment]
@jajodiaraghav will update it by today

**Comment by djmgit** [comment]
Closing this as of now, will be creating a new PR.

**Pull request closed without merging**